### PR TITLE
Elastic Cloud node roles support

### LINKED ISSUE #6950: /dev by zero with Elastic Cloud
Hi All,

i'm using trino db with the elasticsearch plugin on Elastic Cloud managed byAzure
I'm not have problem with config , the plugin works and I'm able to connect with ES instance and
see alla catalogs and schemas and table but , when I'm trying to execute a simple query on any
table , it fails whit the error :

**rg.jkiss.dbeaver.model.sql.DBSQLException: SQL Error [65536]: Query failed (#20210218_151706_00047_xr6qj): / by zero
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.executeStatement(JDBCStatementImpl.java:133)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.executeStatement(SQLQueryJob.java:509)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.lambda$0(SQLQueryJob.java:440)
	at org.jkiss.dbeaver.model.exec.DBExecUtils.tryExecuteRecover(DBExecUtils.java:168)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.executeSingleQuery(SQLQueryJob.java:427)
	at org.jkiss.dbeaver.ui.editors.sql.execute.SQLQueryJob.extractData(SQLQueryJob.java:812)
	at org.jkiss.dbeaver.ui.editors.sql.SQLEditor$QueryResultsContainer.readData(SQLEditor.java:3220)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.lambda$0(ResultSetJobDataRead.java:121)
	at org.jkiss.dbeaver.model.exec.DBExecUtils.tryExecuteRecover(DBExecUtils.java:168)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetJobDataRead.run(ResultSetJobDataRead.java:119)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetViewer$ResultSetDataPumpJob.run(ResultSetViewer.java:4516)
	at org.jkiss.dbeaver.model.runtime.AbstractJob.run(AbstractJob.java:105)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.sql.SQLException: Query failed (#20210218_151706_00047_xr6qj): / by zero
	at io.trino.jdbc.AbstractTrinoResultSet.resultsException(AbstractTrinoResultSet.java:1912)
	at io.trino.jdbc.TrinoResultSet.getColumns(TrinoResultSet.java:242)
	at io.trino.jdbc.TrinoResultSet.create(TrinoResultSet.java:53)
	at io.trino.jdbc.TrinoStatement.internalExecute(TrinoStatement.java:249)
	at io.trino.jdbc.TrinoStatement.execute(TrinoStatement.java:227)
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.execute(JDBCStatementImpl.java:327)
	at org.jkiss.dbeaver.model.impl.jdbc.exec.JDBCStatementImpl.executeStatement(JDBCStatementImpl.java:130)
	... 12 more
Caused by: java.lang.ArithmeticException: / by zero
	at io.trino.plugin.elasticsearch.client.ElasticsearchClient.getSearchShards(ElasticsearchClient.java:425)
	at io.trino.plugin.elasticsearch.ElasticsearchSplitManager.getSplits(ElasticsearchSplitManager.java:60)
	at io.trino.split.SplitManager.getSplits(SplitManager.java:87)
	at io.trino.sql.planner.DistributedExecutionPlanner$Visitor.visitScanAndFilter(DistributedExecutionPlanner.java:203)
	at io.trino.sql.planner.DistributedExecutionPlanner$Visitor.visitTableScan(DistributedExecutionPlanner.java:185)
	at io.trino.sql.planner.DistributedExecutionPlanner$Visitor.visitTableScan(DistributedExecutionPlanner.java:156)
	at io.trino.sql.planner.plan.TableScanNode.accept(TableScanNode.java:143)
	at io.trino.sql.planner.DistributedExecutionPlanner.doPlan(DistributedExecutionPlanner.java:124)
	at io.trino.sql.planner.DistributedExecutionPlanner.doPlan(DistributedExecutionPlanner.java:131)
	at io.trino.sql.planner.DistributedExecutionPlanner.plan(DistributedExecutionPlanner.java:101)
	at io.trino.execution.SqlQueryExecution.planDistribution(SqlQueryExecution.java:470)
	at io.trino.execution.SqlQueryExecution.start(SqlQueryExecution.java:386)
	at io.trino.execution.SqlQueryManager.createQuery(SqlQueryManager.java:237)
	at io.trino.dispatcher.LocalDispatchQuery.lambda$startExecution$7(LocalDispatchQuery.java:143)
	at io.trino.$gen.Trino_351____20210218_115835_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)**

I'm not understanding exactly why. I've found the problem is here, in class ElasticsearchClient.java at the method 

public List<Shard> getSearchShards(String index) 

            SearchShardsResponse.Shard chosen;
            ElasticsearchNode node;
            if (candidate.isEmpty()) {
                // pick an arbitrary shard with and assign to an arbitrary node
                chosen = shardGroup.stream()
                        .min(this::shardPreference)
                        .get();
                **### node = nodes.get(chosen.getShard() % nodes.size());**
            }
            else {
                chosen = candidate.get();
                node = nodeById.get(chosen.getNode());
            }

   I'm also noted that in my system.nodes , table nodes is empty. can means anything?

Thanks in advance

Piero

### PR DESCRIPTION
**Pull request by dingdayu**
Fix: Query failed (#20211107_155438_00000_nmyc7): / by zero

In recent versions, Elastic Cloud uses roles to distinguish data node roles, removing the original data role.

Reference #7017
Fixes #6950

### CODE REVIEW

**Review by martint — APPROVED** [review]
Thanks. The change looks good. Can you please sign the CLA so we can merge it?

Also, please take a look at the checkstyle failure:

> Error:  src/main/java/io/trino/plugin/elasticsearch/client/ElasticsearchClient.java:[132,104] (whitespace) SingleSpaceSeparator: Use a single space to separate non-whitespace characters.

**Comment by aalbu** [comment]
LGTM.  Please fix the checkstyle warning (extra whitespace before `"data_warm"`).

**Comment by dingdayu** [comment]
> LGTM. Please fix the checkstyle warning (extra whitespace before `"data_warm"`).

@aalbu 

Thanks a lot for the reminder, I have fixed it.

I have sent the "Trino Foundation Individual CLA" via email, I don't know if there is an error in the operation, please help me.

**Comment by martint** [comment]
> I have sent the "Trino Foundation Individual CLA" via email, I don't know if there is an error in the operation, please help me.

We haven't received the email. Can you try sending it again?

**Review by hashhar — APPROVED** [review]
LGTM.

**Comment by dingdayu** [comment]
> > I have sent the "Trino Foundation Individual CLA" via email, I don't know if there is an error in the operation, please help me.
> 
> We haven't received the email. Can you try sending it again?

Ok, I will send again

**Review by wendigo — APPROVED** [review]

**Comment by dingdayu** [comment]
@martint I sent CLA email, can you help me confirm it?

**Comment by stevenayers** [comment]
Any update on the CLA? Would love to see this merged in ASAP :)

**Comment by hashhar** [comment]
@cla-bot check

**Comment by martint** [comment]
@cla-bot check

File: plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/client/ElasticsearchClient.java [hunk 1]
```diff
@@ -128,6 +129,7 @@
     private static final ObjectMapper OBJECT_MAPPER = new ObjectMapperProvider().get();
 
     private static final Pattern ADDRESS_PATTERN = Pattern.compile("((?<cname>[^/]+)/)?(?<ip>.+):(?<port>\\d+)");
+    private static final Set<String> NODE_ROLES = ImmutableSet.of("data", "data_content", "data_hot", "data_warm", "data_cold", "data_frozen");
```
Comment (ebyhr) [inline plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/client/ElasticsearchClient.java:132]: How about moving to config property? Any reason we would to do hard-code these names?

**Pull request merged**